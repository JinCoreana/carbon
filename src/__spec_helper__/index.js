import Enzyme from "enzyme";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import { setup } from "./mock-match-media";
import setupResizeObserverMock from "./mock-resize-observer";

require("jest-fetch-mock").enableMocks();

setupResizeObserverMock();
setup();
Enzyme.configure({ adapter: new Adapter() });
