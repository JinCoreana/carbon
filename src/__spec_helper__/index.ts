import { enableFetchMocks } from "jest-fetch-mock";
import { setupMatchMediaMock } from "./mock-match-media";
import setupResizeObserverMock from "./mock-resize-observer";

enableFetchMocks();
setupResizeObserverMock();
setupMatchMediaMock();
