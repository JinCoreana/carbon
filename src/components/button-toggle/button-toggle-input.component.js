import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import {
  StyledButtonToggleInput
} from './button-toggle.style';

import { InputGroupContext } from '../../__internal__/input-behaviour';

function ButtonToggleInput(props) {
  const { onFocus, onBlur } = useContext(InputGroupContext);

  const handleBlur = (ev) => {
    if (props.onBlur) { props.onBlur(ev); }
    if (onBlur) onBlur(ev);
  };

  return (
    <StyledButtonToggleInput
      type='radio'
      name={ props.name }
      id={ props.guid }
      disabled={ props.disabled }
      checked={ props.checked }
      onChange={ props.onChange }
      onBlur={ handleBlur }
      onFocus={ onFocus }
      value={ props.value }
    />
  );
}

ButtonToggleInput.propTypes = {
  /** Set the checked value of the radio button */
  checked: PropTypes.bool,
  /** Name used on the hidden radio button. */
  name: PropTypes.string,
  /** Disable all user interaction. */
  disabled: PropTypes.bool,
  /** Unique ID attribute for input */
  guid: PropTypes.string,
  /** Value for the input */
  value: PropTypes.string,
  /** Callback triggered by change event on the input. */
  onChange: PropTypes.func,
  /** Callback triggered by blur event on the input. */
  onBlur: PropTypes.func
};

export default ButtonToggleInput;
