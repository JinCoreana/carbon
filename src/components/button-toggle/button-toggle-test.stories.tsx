import React from "react";
import { action } from "@storybook/addon-actions";

import ButtonToggle from ".";

export default {
  title: "Button Toggle/Test",
  includeStories: ["DefaultStory"],
  parameters: {
    info: { disable: true },
    chromatic: {
      disableSnapshot: true,
    },
  },
};

export const DefaultStory = () => (
  <div>
    <ButtonToggle
      name="new-button-toggle"
      key="button-toggle-1"
      onChange={() => action("onChange")("foo")}
      onFocus={() => action("onFocus")("foo")}
      onBlur={() => action("onBlur")("foo")}
    >
      Foo
    </ButtonToggle>
    <ButtonToggle
      name="new-button-toggle"
      key="button-toggle-2"
      onChange={() => action("onChange")("bar")}
      onFocus={() => action("onFocus")("bar")}
      onBlur={() => action("onBlur")("bar")}
    >
      Bar
    </ButtonToggle>
    <ButtonToggle
      name="new-button-toggle"
      key="button-toggle-3"
      onChange={() => action("onChange")("baz")}
      onFocus={() => action("onFocus")("baz")}
      onBlur={() => action("onBlur")("baz")}
    >
      Baz
    </ButtonToggle>
  </div>
);

DefaultStory.storyName = "default";

export const ButtonToggleComponent = ({
  // eslint-disable-next-line react/prop-types
  children = "This is an example of an alert",
  ...props
}) => {
  return (
    <div>
      <ButtonToggle
        name="button-toggle-one"
        onBlur={function noRefCheck() {
          ("");
        }}
        onChange={function noRefCheck() {
          ("");
        }}
        onFocus={function noRefCheck() {
          ("");
        }}
        {...props}
      >
        {children}
      </ButtonToggle>
      <ButtonToggle
        name="button-toggle-two"
        onBlur={function noRefCheck() {
          ("");
        }}
        onChange={function noRefCheck() {
          ("");
        }}
        onFocus={function noRefCheck() {
          ("");
        }}
        {...props}
      >
        Second
      </ButtonToggle>
      <ButtonToggle
        name="button-toggle-three"
        onBlur={function noRefCheck() {
          ("");
        }}
        onChange={function noRefCheck() {
          ("");
        }}
        onFocus={function noRefCheck() {
          ("");
        }}
        {...props}
      >
        Third
      </ButtonToggle>
    </div>
  );
};
