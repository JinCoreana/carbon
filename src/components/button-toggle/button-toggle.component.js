import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import {
  StyledButtonToggle,
  StyledButtonToggleLabel,
  StyledButtonToggleContentWrapper
} from './button-toggle.style';
import guid from '../../utils/helpers/guid';
import ButtonToggleIcon from './button-toggle-icon.component';
import ButtonToggleInput from './button-toggle-input.component';
import OptionsHelper from '../../utils/helpers/options-helper';

import { InputGroupContext } from '../../__internal__/input-behaviour';

const ButtonToggle = (props) => {
  const {
    name,
    checked,
    grouped,
    children,
    disabled,
    buttonIcon,
    buttonIconSize,
    onChange,
    onBlur,
    value,
    size
  } = props;
  const { onMouseEnter, onMouseLeave } = useContext(InputGroupContext);

  const inputGuid = guid();
  let icon;

  if (buttonIcon) {
    icon = (
      <ButtonToggleIcon
        buttonIcon={ buttonIcon }
        buttonIconSize={ buttonIconSize }
        disabled={ disabled }
      />
    );
  }

  return (
    <StyledButtonToggle
      data-component='button-toggle'
      grouped={ grouped }
    >
      <ButtonToggleInput
        name={ name }
        checked={ checked }
        disabled={ disabled }
        guid={ inputGuid }
        value={ value }
        onChange={ onChange }
        onBlur={ onBlur }
      />
      <StyledButtonToggleLabel
        buttonIcon={ buttonIcon }
        buttonIconSize={ buttonIconSize }
        disabled={ disabled }
        htmlFor={ inputGuid }
        size={ size }
        onMouseEnter={ onMouseEnter }
        onMouseLeave={ onMouseLeave }
      >
        <StyledButtonToggleContentWrapper>
          { icon }
          { children }
        </StyledButtonToggleContentWrapper>
      </StyledButtonToggleLabel>
    </StyledButtonToggle>
  );
};

ButtonToggle.propTypes = {
  /** Set the checked value of the radio button */
  checked: PropTypes.bool,
  /** Name used on the hidden radio button. */
  name: PropTypes.string,
  /** Callback triggered by change event on the input. */
  onChange: PropTypes.func,
  /** Callback triggered by blur event on the input. */
  onBlur: PropTypes.func,
  /** buttonIcon to render. */
  buttonIcon: PropTypes.string,
  /** Sets the size of the buttonIcon (eg. large) */
  buttonIconSize: PropTypes.oneOf(OptionsHelper.sizesBinary),
  /** Sets the size of the button (padding, font-size). Only used in Classic theme. */
  size: PropTypes.string,
  /** Remove spacing from between buttons. */
  grouped: PropTypes.bool,
  /** Disable all user interaction. */
  disabled: PropTypes.bool,
  /** A required prop. This is the button text. */
  children: PropTypes.node.isRequired,
  /** Set the default value of the Group if component is meant to be used as uncontrolled. */
  defaultChecked: PropTypes.bool, // eslint-disable-line react/no-unused-prop-types
  /** Value for the input */
  value: PropTypes.string
};

ButtonToggle.defaultProps = {
  size: 'large',
  buttonIconSize: 'small'
};

export default ButtonToggle;
