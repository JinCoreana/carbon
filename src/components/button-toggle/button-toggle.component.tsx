import React, { useContext, useRef } from "react";
import invariant from "invariant";
import {
  StyledButtonToggle,
  StyledButtonToggleLabel,
  StyledButtonToggleContentWrapper,
  StyledButtonToggleLabelProps,
} from "./button-toggle.style";
import guid from "../../__internal__/utils/helpers/guid";
import ButtonToggleIcon from "./button-toggle-icon.component";
import ButtonToggleInput, {
  ButtonToggleInputProps,
} from "./button-toggle-input.component";

import { InputGroupContext } from "../../__internal__/input-behaviour";

import Logger from "../../__internal__/utils/logger";

let deprecateUncontrolledWarnTriggered = false;

export interface ButtonToggleProps
  extends ButtonToggleInputProps,
    Partial<StyledButtonToggleLabelProps> {
  /** Text to display for the button. */
  children?: React.ReactNode;
  /** Identifier used for testing purposes, applied to the root element of the component. */
  "data-component"?: string;
  /** Identifier used for testing purposes, applied to the root element of the component. */
  "data-element"?: string;
  /** Identifier used for testing purposes, applied to the root element of the component. */
  "data-role"?: string;
  /** Set the default value of the Group if component is meant to be used as uncontrolled. */
  defaultChecked?: boolean;
  /** Remove spacing from between buttons. */
  grouped?: boolean;
  /** Callback triggered by click event on the input. */
  onClick?: (ev: React.MouseEvent<HTMLInputElement>) => void;
}

export const ButtonToggle = ({
  "aria-label": ariaLabel,
  "aria-labelledby": ariaLabelledBy,
  buttonIcon,
  buttonIconSize = "small",
  checked,
  children,
  "data-component": dataComponent,
  "data-element": dataElement,
  "data-role": dataRole,
  disabled,
  grouped,
  name,
  onBlur,
  onChange,
  onFocus,
  size = "medium",
  value,
}: ButtonToggleProps) => {
  invariant(
    !!(children || buttonIcon),
    "Either prop `buttonIcon` must be defined, or this node must have children"
  );

  const { onMouseEnter, onMouseLeave } = useContext(InputGroupContext);

  const inputGuid = guid();
  const inputRef = useRef<HTMLInputElement>(null);
  let icon;

  if (buttonIcon) {
    icon = (
      <ButtonToggleIcon
        buttonIcon={buttonIcon}
        buttonIconSize={buttonIconSize}
        disabled={disabled}
        hasContent={!!children}
      />
    );
  }

  function handleClick() {
    inputRef.current?.focus();
  }

  if (!deprecateUncontrolledWarnTriggered && !onChange) {
    deprecateUncontrolledWarnTriggered = true;
    Logger.deprecate(
      "Uncontrolled behaviour in `Button Toggle` is deprecated and support will soon be removed. Please make sure all your inputs are controlled."
    );
  }

  return (
    <StyledButtonToggle
      data-component={dataComponent || "button-toggle"}
      data-element={dataElement}
      data-role={dataRole}
      grouped={grouped}
      onClick={handleClick}
    >
      <ButtonToggleInput
        aria-label={ariaLabel}
        aria-labelledby={ariaLabelledBy}
        data-element="button-toggle-input"
        name={name}
        checked={checked}
        disabled={disabled}
        guid={inputGuid}
        value={value}
        onChange={onChange}
        onFocus={onFocus}
        onBlur={onBlur}
        ref={inputRef}
      />
      <StyledButtonToggleLabel
        buttonIcon={buttonIcon}
        buttonIconSize={buttonIconSize}
        disabled={disabled}
        htmlFor={inputGuid}
        onMouseEnter={onMouseEnter}
        onMouseLeave={onMouseLeave}
        size={size}
      >
        <StyledButtonToggleContentWrapper>
          {icon}
          {children}
        </StyledButtonToggleContentWrapper>
      </StyledButtonToggleLabel>
    </StyledButtonToggle>
  );
};

ButtonToggle.displayName = "ButtonToggle";
export default ButtonToggle;
