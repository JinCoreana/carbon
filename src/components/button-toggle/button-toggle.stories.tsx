import React from "react";
import { ComponentStory } from "@storybook/react";

import ButtonToggle from ".";
import ButtonToggleGroup from "../button-toggle-group";
import Box from "../box";

export const Default: ComponentStory<typeof ButtonToggle> = () => (
  <ButtonToggleGroup id="button-toggle-group-id" name="options">
    <ButtonToggle name="default" value="foo">
      Foo
    </ButtonToggle>
    <ButtonToggle name="default" value="bar">
      Bar
    </ButtonToggle>
    <ButtonToggle name="default" value="baz">
      Baz
    </ButtonToggle>
  </ButtonToggleGroup>
);

export const DefaultWrappedText: ComponentStory<typeof ButtonToggle> = () => (
  <Box margin={4} width="250px" display="flex" flexWrap="nowrap">
    <ButtonToggle name="default" value="wraps" grouped>
      Some text that wraps
    </ButtonToggle>
    <ButtonToggle name="default" value="foobar" grouped>
      FooBar
    </ButtonToggle>
  </Box>
);

export const DefaultSmallIcon: ComponentStory<typeof ButtonToggle> = () => (
  <ButtonToggleGroup id="button-toggle-group-id" name="options">
    <ButtonToggle name="sizes" value="foo" buttonIcon="add">
      Add
    </ButtonToggle>
    <ButtonToggle name="sizes" value="bar" buttonIcon="share">
      Share
    </ButtonToggle>
    <ButtonToggle name="sizes" value="baz" buttonIcon="tick">
      Tick
    </ButtonToggle>
  </ButtonToggleGroup>
);

export const DefaultLargeIcon: ComponentStory<typeof ButtonToggle> = () => (
  <ButtonToggleGroup id="button-toggle-group-id" name="options">
    <ButtonToggle
      name="sizes"
      value="foo"
      buttonIcon="add"
      buttonIconSize="large"
    >
      Add
    </ButtonToggle>
    <ButtonToggle
      name="sizes"
      value="bar"
      buttonIcon="share"
      buttonIconSize="large"
    >
      Share
    </ButtonToggle>
    <ButtonToggle
      name="sizes"
      value="baz"
      buttonIcon="tick"
      buttonIconSize="large"
    >
      Tick
    </ButtonToggle>
  </ButtonToggleGroup>
);

export const iconOnly: ComponentStory<typeof ButtonToggle> = () => (
  <ButtonToggleGroup id="button-toggle-group-id" name="options">
    <ButtonToggle name="sizes" value="foo" buttonIcon="add" aria-label="add" />
    <ButtonToggle
      name="sizes"
      value="bar"
      buttonIcon="share"
      aria-label="share"
    />
    <ButtonToggle
      name="sizes"
      value="baz"
      buttonIcon="tick"
      aria-label="tick"
    />
  </ButtonToggleGroup>
);

export const small: ComponentStory<typeof ButtonToggle> = () => (
  <ButtonToggleGroup id="button-toggle-group-id" name="options">
    <ButtonToggle size="small" name="sizes" value="foo">
      Add
    </ButtonToggle>
    <ButtonToggle size="small" name="sizes" value="bar">
      Share
    </ButtonToggle>
    <ButtonToggle size="small" name="sizes" value="baz">
      Tick
    </ButtonToggle>
  </ButtonToggleGroup>
);

export const smallSmallIcon: ComponentStory<typeof ButtonToggle> = () => (
  <ButtonToggleGroup id="button-toggle-group-id" name="options">
    <ButtonToggle size="small" name="sizes" value="foo" buttonIcon="add">
      Add
    </ButtonToggle>
    <ButtonToggle size="small" name="sizes" value="bar" buttonIcon="share">
      Share
    </ButtonToggle>
    <ButtonToggle size="small" name="sizes" value="baz" buttonIcon="tick">
      Tick
    </ButtonToggle>
  </ButtonToggleGroup>
);

export const smallLargeIcon: ComponentStory<typeof ButtonToggle> = () => (
  <ButtonToggleGroup id="button-toggle-group-id" name="options">
    <ButtonToggle
      size="small"
      name="sizes"
      value="foo"
      buttonIcon="add"
      buttonIconSize="large"
    >
      Add
    </ButtonToggle>
    <ButtonToggle
      size="small"
      name="sizes"
      value="bar"
      buttonIcon="share"
      buttonIconSize="large"
    >
      Share
    </ButtonToggle>
    <ButtonToggle
      size="small"
      name="sizes"
      value="baz"
      buttonIcon="tick"
      buttonIconSize="large"
    >
      Tick
    </ButtonToggle>
  </ButtonToggleGroup>
);

export const large: ComponentStory<typeof ButtonToggle> = () => (
  <ButtonToggleGroup id="button-toggle-group-id" name="options">
    <ButtonToggle size="large" name="sizes" value="foo">
      Add
    </ButtonToggle>
    <ButtonToggle size="large" name="sizes" value="bar">
      Share
    </ButtonToggle>
    <ButtonToggle size="large" name="sizes" value="baz">
      Tick
    </ButtonToggle>
  </ButtonToggleGroup>
);

export const largeSmallIcon: ComponentStory<typeof ButtonToggle> = () => (
  <ButtonToggleGroup id="button-toggle-group-id" name="options">
    <ButtonToggle size="large" name="sizes" value="foo" buttonIcon="add">
      Add
    </ButtonToggle>
    <ButtonToggle size="large" name="sizes" value="bar" buttonIcon="share">
      Share
    </ButtonToggle>
    <ButtonToggle size="large" name="sizes" value="baz" buttonIcon="tick">
      Tick
    </ButtonToggle>
  </ButtonToggleGroup>
);

export const largeLargeIcon: ComponentStory<typeof ButtonToggle> = () => (
  <ButtonToggleGroup id="button-toggle-group-id" name="options">
    <ButtonToggle
      size="large"
      name="sizes"
      value="foo"
      buttonIcon="add"
      buttonIconSize="large"
    >
      Add
    </ButtonToggle>
    <ButtonToggle
      size="large"
      name="sizes"
      value="bar"
      buttonIcon="share"
      buttonIconSize="large"
    >
      Share
    </ButtonToggle>
    <ButtonToggle
      size="large"
      name="sizes"
      value="baz"
      buttonIcon="tick"
      buttonIconSize="large"
    >
      Tick
    </ButtonToggle>
  </ButtonToggleGroup>
);

export const disabled: ComponentStory<typeof ButtonToggle> = () => (
  <ButtonToggleGroup id="button-toggle-group-id" name="options">
    <ButtonToggle name="disabled" value="foo" disabled>
      Foo
    </ButtonToggle>
    <ButtonToggle name="disabled" value="bar" disabled>
      Bar
    </ButtonToggle>
    <ButtonToggle name="disabled" value="baz" disabled>
      Baz
    </ButtonToggle>
  </ButtonToggleGroup>
);

export const grouped: ComponentStory<typeof ButtonToggle> = () => (
  <ButtonToggleGroup id="button-toggle-group-id" name="grouped-options">
    <ButtonToggle name="grouped" value="foo" grouped>
      Foo
    </ButtonToggle>
    <ButtonToggle name="grouped" value="bar" grouped>
      Bar
    </ButtonToggle>
    <ButtonToggle name="grouped" value="baz" grouped>
      Baz
    </ButtonToggle>
  </ButtonToggleGroup>
);
