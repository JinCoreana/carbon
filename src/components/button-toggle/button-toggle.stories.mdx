import { useState } from "react";
import { Meta, Story, Canvas, ArgsTable } from "@storybook/addon-docs";
import LinkTo from "@storybook/addon-links/react";

import ButtonToggle from ".";
import ButtonToggleGroup from "../button-toggle-group";

import * as stories from "./button-toggle.stories";

<Meta 
  title="Button Toggle"
  parameters={{
    info: { disable: true },
    themeProvider: { chromatic: { theme: "sage" } },
  }}
/>

# Button Toggle

<a
  target="_blank"
  href="https://zeroheight.com/2ccf2b601/p/996a28-button-toggle/b/7384f9"
  style={{ color: '#007E45', fontWeight: 'bold', textDecoration: 'underline' }}
>
  Product Design System component
</a>

Press one of the buttons to make selection. This component should be used when user has to make a choice between a small number of options.

## Contents

- [Quick Start](#quick-start)
- [Examples](#examples)
- [Props](#props)

## Quick Start

```javascript
import ButtonToggle from "carbon-react/lib/components/button-toggle";
import ButtonToggleGroup from "carbon-react/lib/components/button-toggle-group";
```

- You should use the <LinkTo kind="Button Toggle Group" story="default">ButtonToggleGroup component</LinkTo> if you need to wrap multiple `ButtonToggle` components together.
- An icon could be added to make the meaning of an option clearer.

## Examples

### Default

<Canvas>
  <Story name="default" story={stories.Default} />
</Canvas>

### Default with wrapped text

<Canvas>
  <Story name="default with wrapped text" story={stories.DefaultWrappedText} />
</Canvas>

### Default with small icon

<Canvas>
  <Story name="default with small icon" story={stories.DefaultSmallIcon} />
</Canvas>

### Default with large icon

<Canvas>
  <Story name="default with large icon" story={stories.DefaultLargeIcon} />
</Canvas>

### Icon only

While we do not advocate the use of icon-only buttons for accessibility reasons, the `aria-label` or `aria-labelledby` props can be used with `buttonIcon` so assistive techniques are aware of the action that will be taken when the component pressed.

<Canvas>
  <Story name="icon only" story={stories.iconOnly} />
</Canvas>

### Small

<Canvas>
  <Story name="small" story={stories.small} />
</Canvas>

### Small with small icon

<Canvas>
  <Story name="small with small icon" story={stories.smallSmallIcon} />
</Canvas>

### Small with large icon

<Canvas>
  <Story name="small with large icon" story={stories.smallLargeIcon} />
</Canvas>

### Large

<Canvas>
  <Story name="large" story={stories.large} />
</Canvas>

### Large with small icon

<Canvas>
  <Story name="large with small icon" story={stories.largeSmallIcon} />
</Canvas>

### Large with large icon

<Canvas>
  <Story name="large with large icon" story={stories.largeLargeIcon} />
</Canvas>

### Disabled

<Canvas>
  <Story name="disabled" story={stories.disabled} />
</Canvas>

### Grouped

<Canvas>
  <Story name="grouped" story={stories.grouped} />
</Canvas>

## Props

### Button Toggle

<ArgsTable of={ButtonToggle} />
