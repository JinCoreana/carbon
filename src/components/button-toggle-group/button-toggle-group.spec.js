import React from 'react';
import { mount } from 'enzyme';
import TestRenderer from 'react-test-renderer';
import 'jest-styled-components';
import { ThemeProvider } from 'styled-components';
import guid from '../../utils/helpers/guid';
import classicTheme from '../../style/themes/classic';
import baseTheme from '../../style/themes/base';
import mintTheme from '../../style/themes/mint';
import { assertStyleMatch } from '../../__spec_helper__/test-utils';
import { StyledButtonToggleLabel } from '../button-toggle/button-toggle.style';
import StyledValidationIcon from '../validations/validation-icon.style';
import Label from '../../__experimental__/components/label';

import ButtonToggleGroup from './button-toggle-group.component';
import ButtonToggle from '../button-toggle/button-toggle.component';
import ButtonToggleGroupStyle from './button-toggle-group.style';

jest.mock('../../utils/helpers/guid');
guid.mockImplementation(() => 'guid-12345');

describe('ButtonToggleGroup', () => {
  describe('Classic theme', () => {
    it('renders correctly with default settings', () => {
      const wrapper = renderWithTheme({ theme: classicTheme });
      expect(wrapper).toMatchSnapshot();
    });
  });
  describe('Modern theme', () => {
    it('renders correctly with default settings', () => {
      const wrapper = renderWithTheme({ theme: mintTheme });
      expect(wrapper).toMatchSnapshot();
    });
  });
  describe('Style props', () => {
    it('renders with the correct width', () => {
      const wrapper = renderWithTheme({ theme: classicTheme, labelInline: true, inputWidth: 48 }, mount);
      assertStyleMatch({
        width: '48%'
      }, wrapper.find(ButtonToggleGroupStyle));
    });
  });

  describe('validations', () => {
    const validationTypes = ['error', 'warning', 'info'];
    describe.each(validationTypes)('when %s prop is passed as string', (type) => {
      it('renders proper styles', () => {
        // eslint-disable-next-line max-len
        const boxShadow = type === 'error' ? `inset 1px 1px 0 ${baseTheme.colors.error},inset -1px -1px 0 ${baseTheme.colors.error}` : undefined;
        const wrapper = renderWithTheme({ theme: baseTheme, [type]: 'Message' }, mount);
        assertStyleMatch(
          {
            boxShadow,
            borderColor: baseTheme.colors[type]
          },
          wrapper.find(ButtonToggleGroupStyle),
          { modifier: `${StyledButtonToggleLabel}` }
        );
      });
      it('renders validation icon on input', () => {
        const wrapper = renderWithTheme({ theme: baseTheme, [type]: 'Message' }, mount);
        expect(wrapper.find(ButtonToggleGroupStyle).find(StyledValidationIcon).exists()).toBe(true);
      });
      it('renders validation icon on label when validationOnLabel passed as true', () => {
        const wrapper = renderWithTheme({
          label: 'Label', theme: baseTheme, [type]: 'Message', validationOnLabel: true
        }, mount);
        expect(wrapper.find(Label).find(StyledValidationIcon).exists()).toBe(true);
      });
    });

    describe.each(validationTypes)('when %s prop is passed as boolean', (type) => {
      it('renders proper styles', () => {
        // eslint-disable-next-line max-len
        const boxShadow = type === 'error' ? `inset 1px 1px 0 ${baseTheme.colors.error},inset -1px -1px 0 ${baseTheme.colors.error}` : undefined;
        const wrapper = renderWithTheme({ theme: baseTheme, [type]: true }, mount);
        assertStyleMatch(
          {
            boxShadow,
            borderColor: baseTheme.colors[type]
          },
          wrapper.find(ButtonToggleGroupStyle),
          { modifier: `${StyledButtonToggleLabel}` }
        );
      });

      it('does not render validation icon', () => {
        const wrapper = renderWithTheme({ theme: baseTheme, [type]: true }, mount);
        expect(wrapper.find(StyledValidationIcon).exists()).toBe(false);
      });
    });
  });
});

function renderWithTheme(props = {}, renderer = TestRenderer.create) {
  const { theme, ...componentProps } = props;

  return renderer(
    <ThemeProvider theme={ theme }>
      <ButtonToggleGroup { ...componentProps }>
        <ButtonToggle
          name='button-toggle-group'
          id='foo'
          value='foo'
          { ...componentProps }
        >
          Foo
        </ButtonToggle>
        <ButtonToggle
          name='button-toggle-group'
          id='bar'
          value='bar'
          { ...componentProps }
        >
          Bar
        </ButtonToggle>
      </ButtonToggleGroup>
    </ThemeProvider>
  );
}
