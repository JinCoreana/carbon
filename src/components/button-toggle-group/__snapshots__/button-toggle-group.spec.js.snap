// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonToggleGroup Modern theme renders correctly with default settings 1`] = `
.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  height: 100%;
}

.c6 {
  display: inline-block;
  position: relative;
  height: 38px;
  padding: 0 24px;
  font-weight: 600;
  cursor: pointer;
  border: 1px solid #668592;
  font-size: 14px;
  background-color: #FFFFFF;
}

input:checked ~ .c6 {
  color: #FFFFFF;
  background-color: #00402E;
}

input:checked ~ .c6 .c8 {
  color: #FFFFFF;
}

input:focus ~ .c6 {
  outline: 3px solid #FFB500;
  z-index: 100;
}

.c6:hover {
  background-color: #E6F6F1;
  border-color: #00402E;
}

.c4 {
  display: inline-block;
  vertical-align: middle;
}

.c4:not(:first-of-type) {
  margin-left: 10px;
}

.c5 {
  position: absolute;
  width: 0;
  height: 0;
  opacity: 0;
}

.c1 + .c0 {
  margin-top: 16px;
}

.c2 {
  display: block;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c3 .c9 {
  margin-left: 4px;
}

<div
  className="c0 c1"
>
  <div
    className="c2"
  >
    <div
      className="c3"
      data-component="button-toggle-group"
      role="group"
    >
      <div
        className="c4"
        data-component="button-toggle"
        onClick={[Function]}
      >
        <input
          checked={false}
          className="c5"
          id="guid-12345"
          name="button-toggle-group"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type="radio"
          value="foo"
        />
        <label
          className="c6"
          htmlFor="guid-12345"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="c7"
          >
            Foo
          </div>
        </label>
      </div>
      <div
        className="c4"
        data-component="button-toggle"
        onClick={[Function]}
      >
        <input
          checked={false}
          className="c5"
          id="guid-12345"
          name="button-toggle-group"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type="radio"
          value="bar"
        />
        <label
          className="c6"
          htmlFor="guid-12345"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="c7"
          >
            Bar
          </div>
        </label>
      </div>
    </div>
  </div>
</div>
`;
